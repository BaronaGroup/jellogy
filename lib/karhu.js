module.exports = require('../build/main/karhu')
