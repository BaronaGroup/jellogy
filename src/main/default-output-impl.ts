export default {
  ERROR: (toLog: any[]) => console.error(toLog), /* tslint:disable-line:no-console */
  WARN: (toLog: any[]) => console.warn(toLog), /* tslint:disable-line:no-console */
  default: (toLog: any[]) => console.log(toLog) /* tslint:disable-line:no-console */
}
