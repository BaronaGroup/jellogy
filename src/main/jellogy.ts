import defaultConfigImpl from '../config/default'
import {Color} from 'ansi-styles'

type Context = string
type LogLevel = string

export interface JellogyConfig {
  logLevels: string[]
  colors: {
    [logLevel: string]: Color | Color[]
  }
  formatBefore: (logLevel: string, context: string, colorANSI: string) => string,
  contextSpecificLogLevels: {
    [context: string]: LogLevel
  }
  defaultLogLevel: LogLevel
  envVariable: string
}

const noColor = {
  open: '',
  close: ''
}

export type LogFunction = (...toLog: any[]) => void

export interface DefaultLogImpl {
  error: LogFunction
  warn: LogFunction
  info: LogFunction
  debug: LogFunction
}

let defaultConfig: JellogyConfig | undefined

if (!defaultConfig) defaultConfig = loadConfig()

function loadConfig(newConfig: JellogyConfig | null = null): JellogyConfig {
  return newConfig || defaultConfigImpl
}

export function configure(config: null | JellogyConfig) {
  defaultConfig = loadConfig(config)
}

export const context = (activeContext: Context) => usingConfig(() => required(defaultConfig)).context(activeContext)

export function usingConfig<LogImpl = DefaultLogImpl>(configSource: JellogyConfig | (() => JellogyConfig)) {
  const config = typeof configSource === 'function' ? configSource() : configSource

  return {
    context: forContext
  }

  function forContext(activeContext: string) {
    const impl: any = {}
    for (const type of config.logLevels) {
      impl[type.toLowerCase()] = (...toLog: any[]) => logEvent(config, activeContext, type, toLog)
    }

    return impl as LogImpl
  }
}

function logEvent(config: JellogyConfig, activeContext: string, logLevel: string, toLog: any[]) {

  if (config.logLevels.indexOf(logLevel) >= config.logLevels.indexOf(getLogLevel(config, activeContext))) return
  const color = config.colors[logLevel] || config.colors.default || noColor
  const openColor = asArray(color).map(c => c.open).join('')
  const before = config.formatBefore(logLevel, activeContext, openColor)
  if (process.stdout.isTTY) {
    const closeColor = asArray(color).reverse().map(c => c.close).join('')
    console.log(before, ...toLog, closeColor) /* tslint:disable-line:no-console */
  } else {
    console.log(before, ...toLog) /* tslint:disable-line:no-console */
  }
}

function getLogLevel(config: JellogyConfig, activeContext: Context) {
  return getOverrideLogLevel(config, activeContext) || config.contextSpecificLogLevels[activeContext] || config.defaultLogLevel
}

function getOverrideLogLevel(config: JellogyConfig, activeContext: Context) {
  return process.env[config.envVariable + '_' + toEnv(activeContext)] || process.env[config.envVariable]
}

function toEnv(activeContext: Context) {
  return activeContext.replace(/[^a-zA-Z0-9]+/g, '_')
}

function required<T>(val: T | undefined): T {
  if (!val) throw new Error('Required value missing')
  return val
}

function asArray<T>(inVal: T | T[]): T[] {
  if (inVal instanceof Array) return inVal
  return [inVal]
}
