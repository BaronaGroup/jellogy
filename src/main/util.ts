import {inspect} from 'util'

export function karhuInspect(obj: any) {
  if (typeof obj === 'string') return obj
  return inspect(obj)
}

export function logLevelsMatch(logLevels1: string[], logLevels2: string[]) {
  const set1 = [...new Set(logLevels1)].sort(),
    set2 = [...new Set(logLevels2)].sort()

  if (set1.length !== set2.length) return false
  for (let i = 0; i < set1.length; ++i) {
    if (set1[i] !== set2[i]) return false
  }
  return true
}
