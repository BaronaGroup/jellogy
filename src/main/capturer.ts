import {KarhuLogger, LogFunction} from './karhu'

let temporarilyDisabled = false

export function toggleForceCaptureDisabled(disabled: boolean) {
  temporarilyDisabled = disabled
}

export function enableStandardOutputCapture(karhu: KarhuLogger, stdoutLogLevel: string, stderrLogLevel: string) {
  const wrappers = [
    wrap(karhu, 'stdout', stdoutLogLevel),
    wrap(karhu, 'stderr', stderrLogLevel),
  ]
  return {
    abort: () => {
      for (const wrapper of wrappers) {
        wrapper.abort()
      }
    }
  }
}

function wrap(karhu: KarhuLogger, stream: 'stdout' | 'stderr', logLevel: string) {
  const original = process[stream].write
  const streamAsAny = process[stream] as any
  streamAsAny.write = (arg1: any, arg2: any, arg3: any): boolean => {
    const data = typeof arg1 === 'string' ? arg1 : arg1.toString('UTF-8')
    const callback = typeof arg2 === 'function' ? arg2 : arg3

    if (temporarilyDisabled || data.length === 0) {
      return original.call(process[stream], arg1, arg2, arg3)
    }
    const logFunction = (karhu as any)[logLevel] as LogFunction
    logFunction(data.endsWith && data.endsWith('\n') ? data.substr(0, data.length - 1) : data)
    if (typeof callback === 'function') {
      callback()
    }
    return true
  }

  return {
    abort: () => {
      process[stream].write = original
    }
  }
}
