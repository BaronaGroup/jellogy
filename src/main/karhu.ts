import defaultConfigImpl from '../config/default'
import {Color} from 'ansi-styles'
import {logLevelsMatch} from './util'

type Context = string
type LogLevel = string

export interface KarhuOutputImpl {
  [logLevel: string]: (toLog: any, logLevel: string, context: string, config: KarhuConfig) => void
}

export interface KarhuConfig {
  logLevels: string[]
  colors: {
    [logLevel: string]: Color | Color[]
  }
  outputFormat: string
  formatters: {
    [outputFormat: string]: (toLog: any[], logLevel: string, context: string, config: KarhuConfig, colorStart: string, colorEnd: string) => any,
  }
  contextSpecificLogLevels: {
    [context: string]: LogLevel
  }
  defaultLogLevel: LogLevel
  envVariablePrefix: string,
  outputMapper: (value: any, logLevel: string, context: string, toLog: any[]) => any,
  outputImpl: KarhuOutputImpl,
  formatNow: (config: KarhuConfig) => string | number
}

const noColor = {
  open: '',
  close: ''
}

export type LogFunction = (...toLog: any[]) => void

export interface KarhuLogger {
  error: LogFunction
  ERROR: LogFunction
  warn: LogFunction
  WARN: LogFunction
  info: LogFunction
  INFO: LogFunction
  debug: LogFunction
  DEBUG: LogFunction
}

let defaultConfig: KarhuConfig | undefined

if (!defaultConfig) defaultConfig = loadConfig()

function loadConfig(newConfig: KarhuConfig | null = null): KarhuConfig {
  const config = newConfig || defaultConfigImpl
  if (!config.formatters[config.outputFormat]) throw new Error('There is no formatter for chosen output format')
  return config
}

export function configure(config: null | KarhuConfig) {
  defaultConfig = loadConfig(config)
}

export interface Karhu<LogImpl> {
  context: (context: string) => LogImpl
  reconfigure: (newConfig: Partial<KarhuConfig>) => void
}

export const context = (activeContext: Context) => usingConfig(() => required(defaultConfig)).context(activeContext)

export function usingConfig<LogImpl = KarhuLogger>(configSource: KarhuConfig | (() => KarhuConfig)): Karhu<LogImpl> {
  const config = typeof configSource === 'function' ? configSource() : configSource

  return {
    context: forContext,
    reconfigure
  }

  function forContext(activeContext: string) {
    const impl: any = {}
    for (const type of config.logLevels) {
      impl[type] = (...toLog: any[]) => logEvent(config, activeContext, type, toLog)
      const typeLower = type.toLowerCase()
      if (!impl[typeLower]) {
        impl[typeLower] = (...toLog: any[]) => logEvent(config, activeContext, type, toLog)
      }
    }

    return impl as LogImpl
  }

  function reconfigure(newConfig: Partial<KarhuConfig>) {
    if (newConfig.logLevels && !logLevelsMatch(config.logLevels, newConfig.logLevels)) {
      throw new Error('Log levels cannot be updated with reconfigure')
    }
    Object.assign(config, newConfig)
  }
}

function logEvent(config: KarhuConfig, activeContext: string, logLevel: string, toLog: any[]) {
  const eventLogPrio = config.logLevels.indexOf(logLevel),
    activeLogLevelPrio = config.logLevels.indexOf(getLogLevel(config, activeContext))

  if (eventLogPrio < activeLogLevelPrio) return

  const
      colorEnabled = isColorEnabled(config),
    color = colorEnabled ? config.colors[logLevel] || config.colors.default || noColor : noColor,
    openColor = asArray(color).map(c => c.open).join(''),
    closeColor = asArray(color).reverse().map(c => c.close).join(''),
    mappedValues = toLog.map(value => config.outputMapper(value, logLevel, activeContext, toLog)),
    outputImpl = config.outputImpl[logLevel] || config.outputImpl.default

  const formatted = config.formatters[config.outputFormat](mappedValues, logLevel, activeContext, config, openColor, closeColor)
  outputImpl(formatted, logLevel, activeContext, config)
}

function getLogLevel(config: KarhuConfig, activeContext: Context) {
  return getOverrideLogLevel(config, activeContext) || config.contextSpecificLogLevels[activeContext] || getOverrideLogLevel(config, null) || config.defaultLogLevel
}

function getOverrideLogLevel(config: KarhuConfig, activeContext: Context | null) {
  const prefix = config.envVariablePrefix + '_LOG_LEVEL'
  return process.env[!activeContext ? prefix : prefix + '_' + toEnv(activeContext)]
}

function toEnv(activeContext: Context) {
  return activeContext.replace(/[^a-zA-Z0-9]+/g, '_')
}

function required<T>(val: T | undefined): T {
  if (!val) throw new Error('Required value missing')
  return val
}

function asArray<T>(inVal: T | T[]): T[] {
  if (inVal instanceof Array) return inVal
  return [inVal]
}

function isColorEnabled(config: KarhuConfig) {
  const override = process.env[config.envVariablePrefix + '_COLOR']
  if (override === '0' || override === 'false') return false
  if (override === '1' || override === 'true') return true
  return process.stdout.isTTY
}
