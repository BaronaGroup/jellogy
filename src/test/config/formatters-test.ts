import {prepareTestKarhu} from '../test-utils'
import {KarhuConfig} from '../../main/karhu'

describe('format-before-test', () => {

  describe('functionality', () => {

    const karhuTest = prepareTestKarhu({
      formatters: {
        banana: (toLog: any[], logLevel: string, context: string, config: KarhuConfig, colorStart: string, colorEnd: string) => {
          return 'Overridden'
        }
      },
      outputFormat: 'banana'
    })

    it('allows replacing the prefix', karhuTest((karhu, output) => {
      karhu.context('test-context').info('Hello')
      expect(output.tracked).toMatchSnapshot()
    }))
  })

  describe('args', () => {
    let args: any

    const karhuTest = prepareTestKarhu({
      formatters: {
        banana: (toLog: any[], logLevel: string, context: string, config: KarhuConfig, colorStart: string, colorEnd: string) => {
          args = [toLog, logLevel, context, colorStart, colorEnd]
          return 'Overridden'
        }
      },
      outputFormat: 'banana'
    })

    it('is called with the right arguments', karhuTest((karhu, output) => {
      karhu.context('test-context').warn('Hello', 1, {a: 1})
      expect(args).toMatchSnapshot()
    }))
  })
})
