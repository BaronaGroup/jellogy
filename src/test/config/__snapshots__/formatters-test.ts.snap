// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format-before-test args is called with the right arguments 1`] = `
Array [
  Array [
    "Hello",
    1,
    Object {
      "a": 1,
    },
  ],
  "WARN",
  "test-context",
  "[93m",
  "[39m",
]
`;

exports[`format-before-test functionality allows replacing the prefix 1`] = `
Map {
  "NONE" => Array [],
  "DEBUG" => Array [],
  "INFO" => Array [
    Array [
      "Overridden",
      "INFO",
      "test-context",
    ],
  ],
  "WARN" => Array [],
  "ERROR" => Array [],
}
`;

exports[`format-before-test transports can provide additional/replacement formatters allows replacing the prefix 1`] = `
Map {
  "NONE" => Array [],
  "DEBUG" => Array [],
  "INFO" => Array [
    Array [
      "Howdy",
      "INFO",
      "test-context",
    ],
  ],
  "WARN" => Array [],
  "ERROR" => Array [],
}
`;
